<a href="https://colab.research.google.com/github/Wayne122/CMPE257/blob/main/Week6_HW4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 6 -- Homework 4

Author: Han-Wei Lin

Dataset source: https://archive.ics.uci.edu/ml/datasets/Heart+Disease

Dataset in-drive ID: 1_2mf4NXzVFyxuZ1QXPHpBNnHcPeGj-H5

Data:
This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to
this date.

##Import library and data

In [ ]:
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
import pandas as pd

auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

downloaded = drive.CreateFile({'id':"1_2mf4NXzVFyxuZ1QXPHpBNnHcPeGj-H5"})
downloaded.GetContentFile('dataset.csv')

df = pd.read_csv('dataset.csv')
df.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

## Apply hyper-parameter optimization to your dataset for Random Forest using BayesianSearchCV

### Create X, y and a train/test split



In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop('target', axis=1)
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)
X_train.shape, y_train.shape, X_test.shape, y_test.shape

X_train.head()

age  sex  cp  trestbps  chol  ...  exang  oldpeak  slope  ca  thal
173   58    1   2       132   224  ...      0      3.2      2   2     3
261   52    1   0       112   230  ...      0      0.0      2   1     2
37    54    1   2       150   232  ...      0      1.6      2   0     3
101   59    1   3       178   270  ...      0      4.2      0   0     3
166   67    1   0       120   229  ...      1      2.6      1   2     3

[5 rows x 13 columns]

### Make a classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier(n_estimators=200, random_state=0)
clf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

### BayesSearchCV

* https://scikit-optimize.github.io/#skopt.BayesSearchCV
* should we use an AUC scorer (not accuracy?)

Note that the following error will be fixed: https://github.com/scikit-learn/scikit-learn/pull/9816
```
label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
```

In [ ]:
!pip install scikit-optimize

import skopt
from skopt import BayesSearchCV

In [ ]:
params = {'n_estimators': (5,200),
          'max_depth': (1, 20)}

optimiser = BayesSearchCV(clf, params, n_jobs=-1, n_iter=50, random_state=0)

optimiser.fit(X_train, y_train)

print(optimiser.score(X_test, y_test), optimiser.best_params_)

/usr/local/lib/python3.6/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "
/usr/local/lib/python3.6/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "
/usr/local/lib/python3.6/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "
/usr/local/lib/python3.6/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "
/usr/local/lib/python3.6/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "
/usr/local/lib/python3.6/dist-packa

0.8289473684210527 OrderedDict([('max_depth', 4), ('n_estimators', 158)])


### Optimized hyper-parameter will be:
- Max tree depth: 4
- Best estimators(trees) quantity: 158

In [ ]:
clf.score(X_test, y_test)

0.8289473684210527